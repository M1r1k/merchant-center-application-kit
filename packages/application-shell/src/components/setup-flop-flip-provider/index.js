export {
  default,
  getFlopflipReconfiguration,
} from './setup-flop-flip-provider';
