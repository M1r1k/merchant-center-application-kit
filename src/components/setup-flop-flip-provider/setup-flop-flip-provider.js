import React from 'react';
import PropTypes from 'prop-types';
import { defaultMemoize } from 'reselect';
import ldAdapter from '@flopflip/launchdarkly-adapter';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import { injectConfiguration } from '@commercetools-local/core/components/configuration';
import FetchUser from '../fetch-user';

class SetupFlopFlip extends React.PureComponent {
  static displayName = 'SetupFlopFlip';
  static propTypes = {
    ldClientSideId: PropTypes.string.isRequired,
    projectKey: PropTypes.string,
    children: PropTypes.node.isRequired,
  };

  createLaunchdarklyAdapterArgs = defaultMemoize(
    (
      clientSideId,
      userId,
      ldTrackingId,
      ldTrackingGroup,
      ldTrackingProject
    ) => ({
      clientSideId,
      user: {
        key: userId,
        custom: {
          id: ldTrackingId,
          group: ldTrackingGroup,
          project: ldTrackingProject,
        },
      },
    })
  );

  render() {
    return (
      <FetchUser>
        {({ isLoading, user }) => (
          <ConfigureFlopFlip
            adapter={ldAdapter}
            adapterArgs={this.createLaunchdarklyAdapterArgs(
              this.props.ldClientSideId,
              user && user.id,
              user && user.launchdarklyTrackingId,
              user && user.launchdarklyTrackingGroup,
              this.props.projectKey
            )}
            shouldDeferAdapterConfiguration={isLoading}
          >
            {/* flop flip only accepts a single child :( */}
            <React.Fragment>{this.props.children}</React.Fragment>
          </ConfigureFlopFlip>
        )}
      </FetchUser>
    );
  }
}

export default injectConfiguration(
  ['tracking', 'ldClientSideId'],
  'ldClientSideId'
)(SetupFlopFlip);
