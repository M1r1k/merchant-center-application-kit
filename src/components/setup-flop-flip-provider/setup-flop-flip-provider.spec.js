import React from 'react';
import { shallow } from 'enzyme';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import { SetupFlopFlipProvider } from './setup-flop-flip-provider';

const createTestProps = props => ({
  ldClientSideId: '111',
  user: {
    id: 'u1',
    launchdarklyTrackingId: '123',
    launchdarklyTrackingGroup: 'ct',
  },
  children: jest.fn(),
  ...props,
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<SetupFlopFlipProvider {...props} />);
  });
  it('should render <ConfigureFlopFlip>', () => {
    expect(wrapper).toRender(ConfigureFlopFlip);
  });
  it('should pass "clientSideId" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        clientSideId: props.ldClientSideId,
      })
    );
  });
  it('should pass "user.key" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          key: props.user.id,
        }),
      })
    );
  });
  it('should pass "user.custom.id" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          custom: expect.objectContaining({
            id: props.user.launchdarklyTrackingId,
          }),
        }),
      })
    );
  });
  it('should pass "user.custom.group" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          custom: expect.objectContaining({
            group: props.user.launchdarklyTrackingGroup,
          }),
        }),
      })
    );
  });
  describe('on initial state', () => {
    it('should pass null for "user.custom.project" as adapter arg', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'adapterArgs',
        expect.objectContaining({
          user: expect.objectContaining({
            custom: expect.objectContaining({
              project: null,
            }),
          }),
        })
      );
    });
  });
  describe('when project key is set', () => {
    beforeEach(() => {
      wrapper.instance().setProjectKey('test-1');
      wrapper.update();
    });
    it('should pass null for "user.custom.project" as adapter arg', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'adapterArgs',
        expect.objectContaining({
          user: expect.objectContaining({
            custom: expect.objectContaining({
              project: 'test-1',
            }),
          }),
        })
      );
    });
  });
  describe('when user is not defined', () => {
    beforeEach(() => {
      props = createTestProps({ user: null });
      wrapper = shallow(<SetupFlopFlipProvider {...props} />);
    });
    it('should set "shouldDeferAdapterConfiguration" to true', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'shouldDeferAdapterConfiguration',
        true
      );
    });
  });
  describe('when user is defined', () => {
    it('should set "shouldDeferAdapterConfiguration" to false', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'shouldDeferAdapterConfiguration',
        false
      );
    });
  });
  it('should render children as function with setProjectKey prop', () => {
    expect(props.children).toHaveBeenCalledWith(
      expect.objectContaining({
        setProjectKey: wrapper.instance().setProjectKey,
      })
    );
  });
});
