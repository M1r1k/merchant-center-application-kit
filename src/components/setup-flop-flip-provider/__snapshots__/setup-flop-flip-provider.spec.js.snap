// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "getFlag": [Function],
      "reconfigure": [Function],
      "updateUserContext": [Function],
    }
  }
  adapterArgs={
    Object {
      "clientSideId": "5979d95f6040390cd07b5e00",
      "user": Object {
        "custom": Object {
          "group": "ct",
          "id": "123",
          "team": Array [
            "abc",
            "def",
          ],
        },
        "key": "u1",
      },
    }
  }
  defaultFlags={
    Object {
      "flagA": true,
    }
  }
  shouldDeferAdapterConfiguration={false}
>
  <div />
</ConfigureFlopflip>
`;
